### ISSUE TITLE
Error when trying to send email using gmail

### ISSUE THREAD

**Issue by doberkofler** [buildbot/buildbot#6451]
I am trying to the `buildbot/buildbot-master:latest` docker image for my CI environment als it works like a charm except when trying to send emails.

I use the following email configuration:

```python
from emailLookup import EmailLookup
mn = reporters.MailNotifier(
	fromaddr='x@x.c',
	sendToInterestedUsers=True,
	extraRecipients=[],
	relayhost='smtp.gmail.com',
	smtpPort=587,
	useTls=True,
	smtpUser=private.auth_username,
	smtpPassword=private.auth_password,
	lookup=EmailLookup(),
	generators=[generator],
)

c['services'].append(mn)
```

and get the following error message:

```
2022-03-11 14:04:03+0000 [ESMTPSender,client] Got exception when handling reporter events
	Traceback (most recent call last):
	  File "/buildbot_venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1750, in gotResult
	    current_context.run(_inlineCallbacks, r, gen, status)
	  File "/buildbot_venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1656, in _inlineCallbacks
	    result = current_context.run(
	  File "/buildbot_venv/lib/python3.9/site-packages/twisted/python/failure.py", line 489, in throwExceptionIntoGenerator
	    return g.throw(self.type, self.value, self.tb)
	  File "/buildbot_venv/lib/python3.9/site-packages/buildbot/reporters/base.py", line 95, in _got_event
	    log.err(e, 'Got exception when handling reporter events')
	--- <exception caught here> ---
	  File "/buildbot_venv/lib/python3.9/site-packages/buildbot/reporters/base.py", line 93, in _got_event
	    yield self.sendMessage(reports)
	twisted.mail._except.TLSRequiredError: 502 Server does not support secure communication via TLS / SSL
	<<< 250-smtp.gmail.com at your service, [62.99.227.122]
	<<< 250-SIZE 35882577
	<<< 250-8BITMIME
	<<< 250-STARTTLS
	<<< 250-ENHANCEDSTATUSCODES
	<<< 250-PIPELINING
	<<< 250-CHUNKING
	<<< 250 SMTPUTF8
```

Any help is appreciated.

**Comment by doberkofler** [comment]
@p12tic I just wanted to also mention that I've already tried the following workaround (described in another SR) and it did not help:

```Dockerfile
RUN apt-get -y install -q openssl && \
	pip3 uninstall -y pyopenssl && \
	pip3 uninstall -y cryptography && \
	pip3 install pyopenssl
```

**Comment by kapil-git1087** [comment]
Getting error "MailNotifier.checkConfig() got an unexpected keyword argument"  'mode'" when using "mode=('change','cancelled','failing','problem','warnings','exception')," in the email notifications

error while parsing config file:
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/twisted/internet/defer.py", line 191, in maybeDeferred
    result = f(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/buildbot/scripts/checkconfig.py", line 57, in checkconfig
    return _loadConfig(basedir=basedir, configFile=configFile, quiet=quiet)
  File "/usr/lib/python3.10/site-packages/buildbot/scripts/checkconfig.py", line 27, in _loadConfig
    config.FileLoader(basedir, configFile).loadConfig()
  File "/usr/lib/python3.10/site-packages/buildbot/config.py", line 151, in loadConfig
    filename, config_dict = loadConfigDict(
--- <exception caught here> ---
  File "/usr/lib/python3.10/site-packages/buildbot/config.py", line 114, in loadConfigDict
    execfile(filename, localDict)
  File "/usr/lib/python3.10/site-packages/twisted/python/compat.py", line 224, in execfile
    exec(code, globals, locals)
  File "/home/linux-patching/buildbot/master.cfg", line 274, in <module>
    mn = reporters.MailNotifier(fromaddr="buildbot@buildbot-test",
  File "/usr/lib/python3.10/site-packages/buildbot/reporters/base.py", line 36, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/lib/python3.10/site-packages/buildbot/util/service.py", line 189, in __init__
    self.checkConfig(*args, **kwargs)
builtins.TypeError: MailNotifier.checkConfig() got an unexpected keyword argument 'mode'

Configuration Errors:
  error while parsing config file: MailNotifier.checkConfig() got an unexpected keyword argument 'mode' (traceback in logfile)

**Comment by Swansky** [comment]
Hello, I still have the same problem. It concerns gmail and outlook. It is more annoying that no fix is proposed.

**Comment by doberkofler** [comment]
@tardyp Still not able to send email via gmail from buildbot master **3.6.0** running in docker container!

Python version: 3.9.2
Buildbot version: 3.6.0
Twisted version: 22.4.0

```
master-config-buildbot-1  | 2022-08-29 05:16:35+0000 [ESMTPSender,client] Got exception when handling reporter events
master-config-buildbot-1  |     Traceback (most recent call last):
master-config-buildbot-1  |       File "/buildbot_venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1750, in gotResult
master-config-buildbot-1  |         current_context.run(_inlineCallbacks, r, gen, status)
master-config-buildbot-1  |       File "/buildbot_venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1656, in _inlineCallbacks
master-config-buildbot-1  |         result = current_context.run(
master-config-buildbot-1  |       File "/buildbot_venv/lib/python3.9/site-packages/twisted/python/failure.py", line 514, in throwExceptionIntoGenerator
master-config-buildbot-1  |         return g.throw(self.type, self.value, self.tb)
master-config-buildbot-1  |       File "/buildbot_venv/lib/python3.9/site-packages/buildbot/reporters/base.py", line 113, in _got_event
master-config-buildbot-1  |         log.err(e, 'Got exception when handling reporter events')
master-config-buildbot-1  |     --- <exception caught here> ---
master-config-buildbot-1  |       File "/buildbot_venv/lib/python3.9/site-packages/buildbot/reporters/base.py", line 111, in _got_event
master-config-buildbot-1  |         yield self.sendMessage(reports)
master-config-buildbot-1  |     twisted.mail._except.TLSRequiredError: 502 Server does not support secure communication via TLS / SSL
```

```python
c['services'] = []

from emailLookup import EmailLookup

generator = reporters.BuildStatusGenerator(
	mode=('change'),
	)

mn = reporters.MailNotifier(
	fromaddr=private.auth_username,
	sendToInterestedUsers=True,
	extraRecipients=[],
	useTls=True,
	relayhost='smtp.gmail.com',
	smtpPort=587,
	smtpUser=private.auth_username,
	smtpPassword=private.auth_password,
	lookup=EmailLookup(),
	generators=[generator],
	)

c['services'].append(mn)
```

**Comment by doberkofler** [comment]
@p12tic Any news on this bug? It is killing me not to receive any emails :-(

**Comment by p12tic** [comment]
@doberkofler I haven't forgot about this, just that I had more pressing priorities for the last half of year. I plan to look into this in the coming 2 weeks or so.

**Comment by doberkofler** [comment]
@p12tic Thank you for caring

**Comment by doberkofler** [comment]
@p12tic I guess you have been able to look into this but maybe you can help me to get my workaround working?

The following code does send the email but unfortunately only once and I guess it might be related to the the return code of the function generated by `defer.Deferred()`. My `new_sendMail` function does not use the `result` variable and maybe it needs to be resolved somehow but I do not understand how to do that?

```python
def sendMail(self, m, recipients):
   s = m.as_string()
   twlog.msg(f"sending mail ({len(s)} bytes) to", recipients)
   if self.dumpMailsToLog:  # pragma: no cover
       twlog.msg(f"mail data:\n{s}")

   result = defer.Deferred()

   recipients = [parseaddr(r)[1] for r in recipients]

   new_sendMail(relayhost=self.relayhost, username=self.smtpUser, password=self.smtpPassword, sendFrom=parseaddr(self.fromaddr)[1], sendTo=recipients, email=s)

   return result
```

**Comment by doberkofler** [comment]
The bug is still preventing any email from being send, but the following version of sendMail in my workaround now seems to do the job:

```python
    def sendMail(self, m, recipients):
        s = m.as_string()
        twlog.msg(f"sending mail ({len(s)} bytes) to", recipients)
        if self.dumpMailsToLog:  # pragma: no cover
            twlog.msg(f"mail data:\n{s}")

        recipients = [parseaddr(r)[1] for r in recipients]

        new_sendMail(relayhost=self.relayhost, username=self.smtpUser, password=self.smtpPassword, sendFrom=parseaddr(self.fromaddr)[1], sendTo=recipients, email=s)

        return defer.succeed(True)
```

**Comment by RarrugasTetra** [comment]
Thank you for sharing this solution! It works also for me. I've already nearly lost the hope to get it working with my sendMail function...

> The bug is still preventing any email from being send, but the following version of sendMail in my workaround now seems to do the job:
> 
> ```
>         return defer.succeed(True)
> ```

**Comment by kapil-git1087** [comment]
Buildbot 3.8 has a bug in the twisted.mail


2023-05-01 16:29:25-0500 [-] Starting factory <twisted.mail.smtp.ESMTPSenderFactory object at 0x7fce491a4c10>
2023-05-01 16:29:25-0500 [ESMTPSender,client] SMTP Client retrying server. Retry: 5
2023-05-01 16:29:25-0500 [ESMTPSender,client] SMTP Client retrying server. Retry: 4
2023-05-01 16:29:25-0500 [ESMTPSender,client] SMTP Client retrying server. Retry: 3
2023-05-01 16:29:26-0500 [ESMTPSender,client] SMTP Client retrying server. Retry: 2
2023-05-01 16:29:26-0500 [ESMTPSender,client] SMTP Client retrying server. Retry: 1
2023-05-01 16:29:26-0500 [ESMTPSender,client] Got exception when handling reporter events
        Traceback (most recent call last):
          File "/usr/lib/python3.11/site-packages/twisted/internet/defer.py", line 1750, in gotResult
            current_context.run(_inlineCallbacks, r, gen, status)
          File "/usr/lib/python3.11/site-packages/twisted/internet/defer.py", line 1656, in _inlineCallbacks
            result = current_context.run(
          File "/usr/lib/python3.11/site-packages/twisted/python/failure.py", line 514, in throwExceptionIntoGenerator
            return g.throw(self.type, self.value, self.tb)
          File "/usr/lib/python3.11/site-packages/buildbot/reporters/base.py", line 113, in _got_event
            log.err(e, 'Got exception when handling reporter events')
        --- <exception caught here> ---
          File "/usr/lib/python3.11/site-packages/buildbot/reporters/base.py", line 111, in _got_event
            yield self.sendMessage(reports)
        twisted.mail._except.SMTPConnectError: Unable to connect to server.

2023-05-01 16:29:26-0500 [-] Stopping factory <twisted.mail.smtp.ESMTPSenderFactory object at 0x7fce491a4c10>

**Comment by doberkofler** [comment]
@p12tic has this now been fixed in 3.10.0 as part of #5609 ?

**Comment by doberkofler** [comment]
Unfortunately it is still not possible to send an email using gmail in 4.0.1

**Comment by doberkofler** [comment]
@p12tic Any news on this bug? It would be great, if this could be fixed!

**Comment by doberkofler** [comment]
@p12tic I think everyone would really appreciate if this very old bug would finally be fixed!

**Comment by muks** [comment]
See the latest comments on #5943 for a patch to test.

**Comment by doberkofler** [comment]
With the latest revision 4.2.0 just removing the `useTls` parameter resolves the problem for me.

```python
mn = reporters.MailNotifier(
	fromaddr=private.auth_username,
	sendToInterestedUsers=True,
	extraRecipients=[],
#	useTls=True,
	relayhost="smtp.gmail.com",
	smtpPort=587,
	smtpUser=private.auth_username,
	smtpPassword=private.auth_password,
	lookup=EmailLookup(),
	generators=[generator],
	)
```
